core/src/test/java/ai/timefold/solver/core/impl/domain/variable/cascade/CascadeUpdateVariableListenerTest.java

```diff
@@ -0,0 +1,168 @@
+import ai.timefold.solver.core.impl.testdata.domain.shadow.wrong_cascade.TestdataCascadeMissingNextValue;
+import ai.timefold.solver.core.impl.testdata.domain.shadow.wrong_cascade.TestdataCascadeMissingPreviousValue;
+import ai.timefold.solver.core.impl.testdata.domain.shadow.wrong_cascade.TestdataCascadeWrongMethod;
+import ai.timefold.solver.core.impl.testdata.util.PlannerTestUtils;
+
+import org.junit.jupiter.api.Test;
+
+class CascadeUpdateVariableListenerTest {
+
+    @Test
+    void requiredShadowVariableDependencies() {
+        assertThatIllegalArgumentException().isThrownBy(TestdataCascadeMissingInverseValue::buildEntityDescriptor)
+                .withMessageContaining(
+                        "The entityClass (class ai.timefold.solver.core.impl.testdata.domain.shadow.wrong_cascade.TestdataCascadeMissingInverseValue)")
+                .withMessageContaining("has an @CascadeUpdateElementShadowVariable annotated property (cascadeValue)")
+                .withMessageContaining("but has no @InverseRelationShadowVariable shadow variable defined.");
+
+        assertThatIllegalArgumentException().isThrownBy(TestdataCascadeMissingPreviousValue::buildEntityDescriptor)
+                .withMessageContaining(
+                        "The entityClass (class ai.timefold.solver.core.impl.testdata.domain.shadow.wrong_cascade.TestdataCascadeMissingPreviousValue)")
+                .withMessageContaining("has an @CascadeUpdateElementShadowVariable annotated property (cascadeValue)")
+                .withMessageContaining("but has no @PreviousElementShadowVariable shadow variable defined");
+
+        assertThatIllegalArgumentException().isThrownBy(TestdataCascadeMissingNextValue::buildEntityDescriptor)
+                .withMessageContaining(
+                        "The entityClass (class ai.timefold.solver.core.impl.testdata.domain.shadow.wrong_cascade.TestdataCascadeMissingNextValue)")
+                .withMessageContaining("has an @CascadeUpdateElementShadowVariable annotated property (cascadeValue)")
+                .withMessageContaining("but has no @NextElementShadowVariable shadow variable defined");
+
+        assertThatIllegalArgumentException().isThrownBy(TestdataCascadeWrongMethod::buildEntityDescriptor)
+                .withMessageContaining(
+                        "The entityClass (class ai.timefold.solver.core.impl.testdata.domain.shadow.wrong_cascade.TestdataCascadeWrongMethod)")
+                .withMessageContaining(
+                        "has an @CascadeUpdateElementShadowVariable annotated property (badUpdateCascadeValueWithReturnType)")
+                .withMessageContaining(
+                        "with sourceMethodName (class ai.timefold.solver.core.impl.testdata.domain.shadow.wrong_cascade.TestdataCascadeWrongMethod), ")
+                .withMessageContaining("but the method has not been found in the entityClass");
+    }
+
+    @Test
+    void updateAllNextValues() {
```

Comment:
Maybe we should also test the case where two listeners point to the same method?

Reply:
Done!